### PR TITLE
Single-element tuples

### PR DESCRIPTION
**Pull request by sornas**

### CODE REVIEW

File: progs/tests/tuples_fail_invalid_singleton.sy [hunk 1]
```diff
@@ -0,0 +1,7 @@
+start :: fn {
+    a := (,)
+    a
+}
+
+//TODO(gu) See #100
+// errors: [ErrorKind::SyntaxError(_, _), ErrorKind::SyntaxError(_, _), ErrorKind::SyntaxError(_, _)]
```
Comment (sornas) [inline progs/tests/tuples_fail_invalid_singleton.sy:7]: #100

File: src/lib.rs [hunk 2]
```diff
@@ -596,6 +601,7 @@ mod op {
             (Value::Int(a), Value::Int(b)) => Value::Int(a + b),
             (Value::String(a), Value::String(b)) => Value::String(Rc::from(format!("{}{}", a, b))),
             (Value::Tuple(a), Value::Tuple(b)) if a.len() == b.len() => tuple_bin_op(a, b, add),
+            (Value::Unknown, a) | (a, Value::Unknown) => a.clone(),
```
Comment (sornas) [inline src/lib.rs:604]: Swallowing a Value::Unknown this way is a bit sketchy. Ideally we'd only swallow when type checking but that's potentially _a lot_ of branching. `#[cfg(debug)]`?

Comment (FredTheDino) [reply src/lib.rs:604]: A nice fix here would be to pass in the tuple (a, a) for checking with this operation, so we don't have to assume the type supports the operation. And then only have the case `(Value::Unknown, Value::Unknown) => Value::Unknown`. Maybe?

File: progs/tests/tuples_fail_invalid_singleton.sy (hunk 1 above)
Comment (FredTheDino) [reply progs/tests/tuples_fail_invalid_singleton.sy:7]: We should remove parse_branch.

**Review by FredTheDino — CHANGES_REQUESTED** [review]
Looks good! But we can do slightly better type inference!

File: progs/tests/tuples_fail_invalid_singleton.sy (hunk 1 above)
Comment (sornas) [reply progs/tests/tuples_fail_invalid_singleton.sy:7]: Well, yes. But I think it is out of scope here.

**Review by FredTheDino — APPROVED** [review]
LGTM

**Pull request merged**